Expressive modal

photodow opened the pull request:
### Related Ticket(s)

closes #253

### Description

This pull request creates a new expressive modal leverages the Carbon's modal component. It mostly adds 32px margin around the whole container, pulls out the close button opening up the ability for content transitions and whatever content needs to go in.

![image](https://user-images.githubusercontent.com/3793636/67624142-ab52fe00-f7f2-11e9-8c7f-f420562c20e7.png)

### Changelog

**New**

- Expressive Modal

**Changed**

- Updated locale button with expressive modal

jeffchew commented on packages/react/src/components/ExpressiveModal/ExpressiveModal.js:
```diff
@@ -0,0 +1,74 @@
+};
+
+ExpressiveModal.defaultProps = {
+  open: false,
```
@photodow I'm wondering if we should also pass in props for the primary and secondary buttons similar to how the productive modal works. They can be optional attributes, and if they don't pass anything then the ButtonGroup doesn't render.

cc: @annawen1 @sannminwin @andysherman2121 @kennylam

photodow replied:
No, I don't think so. The modal itself isn't supposed to have buttons. When you add buttons it becomes a dialogue. The locale selector and lightbox, for example, should use this component, but won't have buttons.

@wonilsuhibm @oliviaflory please correct me if I misunderstood.

oliviaflory replied:
@jeffchew Wonil and I met with James on Friday and decided the expressive modal should just be a shell so adopters can add whatever they need to it. We will still provide the other modals Carbon has in the theming. Sorry for not updating our convo in the original issue.

After a design critique with Carbon on Thursday we think the expressive modal is different than the current modals that exist in Carbon today, which are like a 'dialogue box' that requires an action / decision. Whereas the expressive modal could be used for lightbox or local selector that could have different requirements or no action required. (We need to work on the definition).

jeffchew replied:
Thanks for bringing me up to speed @oliviaflory . It's too bad we can't just call ours `<Modal />` and rename theirs `<Dialog />`

jeffchew commented on packages/react/src/components/ExpressiveModal/ExpressiveModalCloseBtn.js:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Expressive Modal
+ *
+ * @param {object} props react proptypes
```
Can you add JSDocs for the props?

photodow replied:
Sorry, isn't that what this line is?

jeffchew replied:
See example here:
https://github.com/carbon-design-system/ibm-dotcom-library/blob/ebc6b1f4dbf82e0f3eefb8bf53b42bff7992edfe/packages/react/src/components/HorizontalRule/HorizontalRule.js#L31

photodow replied:
At first, glance that looks inaccurate, but I'll dive in and take a closer look. Thanks for the example!

photodow replied:
I see it now, thanks! I'll make the update.

jeffchew commented on packages/react/src/components/ExpressiveModal/ExpressiveModal.js:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Expressive Modal
+ *
+ * @param {object} props react proptypes
```
Can you add JSDocs for the props?

jeffchew commented on packages/react/src/components/ExpressiveModal/ExpressiveModal.js:
```diff
@@ -0,0 +1,74 @@
+   */
+  function manageOpenState() {
+    if (prevOpen !== open) {
+      setIsOpen(open);
```
@photodow one thing I neglected to add in the acceptance criteria are the analytics triggered events in this modal. We could probably do this in a separate PR, but that means we'd have to add it post code freeze if we can't sneak this in today.

photodow replied:
I was assuming that was part of the https://github.com/carbon-design-system/ibm-dotcom-library/issues/391 story. I'm fine with a separate pull request. I don't think code freeze has to be so strict, but it's up to you. At the very least #571 needs to go in first, right?

jeffchew replied:
@photodow Yes, #571 definitely has to go in first. I'm fine with getting this in separately.

oliviaflory commented:
@photodow question about the padding on the expressive modal, is the 30px on the right coming from how Carbon built the modals or us? Carbon's current guidance says text should only span 75% of the modal, is that baked in or do developers have to add that? Only wondering if we should use a percentage vs a fixed pixel size.

<img width="653" alt="Screen Shot 2019-10-29 at 10 18 52 AM" src="https://user-images.githubusercontent.com/45692622/67775783-c9f50700-fa35-11e9-8dab-555118f8493f.png">

photodow commented:
@oliviaflory reached out via slack.

> The spacing on the right clears the close button, would you like the 75%? I removed it because of the designs that had the borders going full across, and y’all didn’t mention keeping it when we talked. It sounded more like a pain or restricting, so I removed it, but if not I can add it back :slightly_smiling_face:

![image](https://user-images.githubusercontent.com/3793636/67779466-ac726e00-fa32-11e9-9bd8-cd2bfdda061d.png)

It's still not perfect because in this design when it scrolls up it'll overlap with the `x`, and the one this pull request has tried to avoid it so there is no overlap. Just let me know, it won't be hard to revert.
